### PR TITLE
OSGi classloading workaround + refactor servlet contants

### LINKED ISSUE #2104: AEM application - traces not showing/capturing outgoing HTTP requests
**Issue Description**

We have currently setup the java apm-agent ( as command argument in JAVA_OPTS) for one of the AEM publisher application. We are able to see #Dispatcherservlet#doGet #Dispatcherservlet#doPost transaction in APM app in Kibana . But we are not able to see any further traces where it has made an outgoing HTTP call to external service .

**Information**
FYI, There is another apm - app dynamics is also installed in the machine and it is capturing all the traces. There are other tomcat servers where we have also configured apm-agent and we are able to see all the outgoing requests

**Version used**
java apm-agent version : 1.25.0
elasticsearch version : 7.14.0
kibana version: 7.14.0
apm-server version : 7.14.0

**Few things we have tried so far**

> Adding properties such as -Delastic.apm.disable_instrumentations=none -Dtrace_methods=* -Delastic.apm.profiling_inferred_spans_enabled=true
> Enabling Trace -Delastic.apm.log_level=TRACE ( not able to see any trace )
> Added sling.bootdelegation.co.elastic.apm=co.elastic.apm.* in sling.properties

Kindly help me by informing what configuration needs to be change to get all the traces. We are planning to replace app-dynamics with elastic-apm but this issue is really blocking us. Please help.

**Error stack trace** - Disclaimer can't give full due to security concern 

**java.lang.NoClassDefFoundError: javax/servlet/http/HttpServlet
        at co.elastic.apm.agent.servlet.ServletApiAdvice.onExitServletService(ServletApiAdvice.java:214)**
        at com.adobe.granite.rest.assets.impl.AssetContentDispositionFilter.doFilter(AssetContentDispositionFilter.java:97)
        at org.apache.sling.engine.impl.filter.AbstractSlingFilterChain.doFilter(AbstractSlingFilterChain.java:72)

### PR DESCRIPTION
**Pull request by SylvainJuge**
~~Add debug logging to help understanding what could cause #2104~~

UPDATE: re-purposing this PR
- avoid loading our Servlet instrumentation with OSGi bundles that only import parts of the `Servlet` API.
- refactor servlet instrumentation constants.
- avoid `SecurityException` thrown by Apache Sling

Fixes #2104

### CODE REVIEW

File: apm-agent-core/src/test/java/co/elastic/apm/agent/logging/LoggingConfigurationTest.java [hunk 1]
```diff
@@ -237,7 +237,7 @@ public LoggerContext getContext(String fqcn, ClassLoader loader, boolean current
                 }
                 LoggerContext loggerContext = contextMap.get(loader);
                 if (loggerContext == null) {
-                    org.apache.logging.log4j.core.LoggerContext loggerContextImpl = new org.apache.logging.log4j.core.LoggerContext(Objects.toString(loader));
+                    org.apache.logging.log4j.core.LoggerContext loggerContextImpl = new org.apache.logging.log4j.core.LoggerContext(loader.getClass().getName());
```
Comment (eyalkoren) [inline apm-agent-core/src/test/java/co/elastic/apm/agent/logging/LoggingConfigurationTest.java:240]:
```suggestion
                    org.apache.logging.log4j.core.LoggerContext loggerContextImpl = new org.apache.logging.log4j.core.LoggerContext(loader.getClass().getName() + String.valueOf(System.identityHashCode(loader)));
```
We want a context per CL instance, not per CL type

**Review by eyalkoren — CHANGES_REQUESTED** [review]

File: apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/Constants.java [hunk 2]
```diff
@@ -18,10 +18,118 @@
...
+        }
+
+        public ElementMatcher.Junction<ClassLoader> getClassloaderFilter() {
+            // async context class has been introduced in servlet spec 3.0
+            // choice of class name to use for this test does not work as expected across all application servers
+            // for example, 'javax.servlet.annotation.WebServlet' annotation is not working as expected on Payara
+            return CustomElementMatchers.classLoaderCanLoadClass(asyncContextClass)
+                // OSGi bundles might only include 'jakarta.servlet' and not 'jakarta.servlet.http' (same for `java.servlet`), hence
+                // this allows to disable instrumentation as it is required for our instrumentation code.
+                .and(CustomElementMatchers.classLoaderCanLoadClass(osgiClassloaderFilterClass));
```
Comment (SylvainJuge) [inline apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/Constants.java:131]: [for reviewer] this is what provides the work-around for OSGi by requiring ability to load a class in `javax.servlet.http.*` or `jakarta.servlet.http.*` packages.

**Comment by SylvainJuge** [comment]
I have also added a simple work-around for Apache Sling throwing `SecurityException`, see https://github.com/elastic/apm-agent-java/issues/2104#issuecomment-1069192135 for context.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/bci/classloading/DiscriminatingMultiParentClassLoader.java [hunk 3]
```diff
@@ -129,4 +129,12 @@ public URL getResource(String name) {
...
         return Collections.emptyEnumeration();
     }
+
+    @Override
+    public String toString() {
+        return "DiscriminatingMultiParentClassLoader{" +
+            "agentClassLoader = " + parents.get(0) + " discriminator = "+ discriminators.get(0) +
+            ", targetClassLoader =" + parents.get(1) + " discriminator = " + discriminators.get(1) +
+            '}';
+    }
```
Comment (eyalkoren) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/bci/classloading/DiscriminatingMultiParentClassLoader.java:139]: 👍

File: apm-agent-core/src/main/java/co/elastic/apm/agent/bci/classloading/ExternalPluginClassLoader.java [hunk 4]
```diff
@@ -41,9 +41,13 @@
  */
 public class ExternalPluginClassLoader extends URLClassLoader {
     private final List<String> classNames;
+    private final File pluginJar;
+    private final ClassLoader agentClassLoader;
 
     public ExternalPluginClassLoader(File pluginJar, ClassLoader agentClassLoader) throws IOException {
         super(new URL[]{pluginJar.toURI().toURL()}, agentClassLoader);
+        this.pluginJar = pluginJar;
+        this.agentClassLoader = agentClassLoader;
```
Comment (eyalkoren) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/bci/classloading/ExternalPluginClassLoader.java:50]: No need, it's accessible through `getParent()`

File: apm-agent-core/src/main/java/co/elastic/apm/agent/bci/classloading/ExternalPluginClassLoader.java [hunk 5]
```diff
@@ -89,4 +93,11 @@ public ExternalPluginClassLoader(File pluginJar, ClassLoader agentClassLoader) t
         return classNames;
     }
 
+    @Override
+    public String toString() {
+        return "ExternalPluginClassLoader{" +
+            "pluginJar=" + pluginJar +
+            ", agentClassLoader=" + agentClassLoader +
```
Comment (eyalkoren) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/bci/classloading/ExternalPluginClassLoader.java:100]: Could that be something other than the single agent CL?

File: apm-agent-core/src/main/java/co/elastic/apm/agent/bci/classloading/IndyPluginClassLoader.java [hunk 6]
```diff
@@ -39,9 +39,14 @@
...
     private static final ClassLoader SYSTEM_CLASS_LOADER = ClassLoader.getSystemClassLoader();
 
+    private final ClassLoader targetClassLoader;
+    private final ClassLoader agentClassLoader;
+
     public IndyPluginClassLoader(@Nullable ClassLoader targetClassLoader, ClassLoader agentClassLoader, Map<String, byte[]> typeDefinitions) {
         // See getResource on why we're using PersistenceHandler.LATENT over PersistenceHandler.MANIFEST
         super(getParent(targetClassLoader, agentClassLoader), true, typeDefinitions, PersistenceHandler.LATENT);
+        this.targetClassLoader = targetClassLoader;
+        this.agentClassLoader = agentClassLoader;
```
Comment (eyalkoren) [inline apm-agent-core/src/main/java/co/elastic/apm/agent/bci/classloading/IndyPluginClassLoader.java:49]: No need for that, only print the value of `getParent().toString()`. 
In addition, implement a proper `toString()` for `ShadedClassLoader`.
Then, you will get what you want either if the parent is the agent CL, or if it is a `DiscriminatingMultiParentClassLoader`.

File: apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/Constants.java (hunk 2 above)
Comment (eyalkoren) [reply apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/Constants.java:131]: This workaround assumes that at least one class on the execution stack is loaded by a class loader that knows both `javax.servlet` and `javax.servlet.http`. So far it seems to be a valid concept, but let's keep that in mind as a possible source for future reports on missing endpoints.

File: apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/Constants.java [hunk 7]
```diff
@@ -18,10 +18,118 @@
...
+
+        public ElementMatcher.Junction<TypeDescription> servletInputStreamClass() {
+            return named(servletInputStreamClass);
+        }
+
+        public ElementMatcher.Junction<TypeDescription> servletConfigClass() {
+            return named(servletConfigClass);
+        }
+
+        public ElementMatcher.Junction<ClassLoader> getClassloaderFilter() {
```
Comment (eyalkoren) [inline apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/Constants.java:124]: [minor] consider ending these accessor names with `Matcher`, for example `servletConfigClassMatcher()`.

File: apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/JakartaAsyncInstrumentation.java [hunk 8]
```diff
@@ -40,30 +40,15 @@
...
-        @Override
-        String asyncContextClassName() {
-            return "jakarta.servlet.AsyncContext";
-        }
-
-        @Override
-        String servletResponseClassName() {
-            return "jakarta.servlet.ServletResponse";
+        public Constants.ServletImpl getImplConstants() {
+            return Constants.ServletImpl.JAKARTA;
```
Comment (eyalkoren) [inline apm-agent-plugins/apm-servlet-plugin/src/main/java/co/elastic/apm/agent/servlet/JakartaAsyncInstrumentation.java:44]: That's nice!

**Review by eyalkoren — COMMENTED** [review]
I really like the refactoring! 👏 
A negative added-removed balance in PR lines is a great indication for something good 😄 

The main comment is about losing class loader references within class loaders. I prefer we only do that on special cases. Class loaders refs within class loaders should represent the delegation hierarchy as much as possible in my opinion.

**Review by eyalkoren — APPROVED** [review]

**Pull request merged**